serde-generate/src/indent.rs

```diff
@@ -0,0 +1,135 @@
+use std::io::{Result, Write};
+
+#[derive(Clone, Copy)]
+pub enum Indent {
+    Tab,
+    Space(usize),
+}
+
+pub struct IndentedWriter<T> {
+    w: T,
+    indent_level: usize,
+    indent: Indent,
+    at_begining_of_line: bool,
+}
+
+impl<T> IndentedWriter<T> {
+    pub fn new(w: T, indent: Indent) -> Self {
+        Self {
+            w,
+            indent_level: 0,
+            indent,
+            at_begining_of_line: true,
+        }
+    }
+
+    pub fn level(&self) -> usize {
+        self.indent_level
+    }
+
+    pub fn set_level(&mut self, level: usize) {
+        self.indent_level = level
+    }
+
+    pub fn inc_level(&mut self) {
+        self.inc_level_by(1);
+    }
+
+    pub fn dec_level(&mut self) {
+        self.dec_level_by(1);
+    }
+
+    pub fn inc_level_by(&mut self, n: usize) {
+        self.indent_level = self.indent_level.saturating_add(n);
+    }
+
+    pub fn dec_level_by(&mut self, n: usize) {
+        self.indent_level = self.indent_level.saturating_sub(n);
+    }
+}
+
+impl<T: Write> Write for IndentedWriter<T> {
+    fn write(&mut self, mut buf: &[u8]) -> Result<usize> {
+        let mut bytes_written = 0;
+
+        while !buf.is_empty() {
+            let (before_newline, newline, after_newline) =
+                if let Some(idx) = buf.iter().position(|&b| b == b'\n') {
+                    (&buf[..idx], true, &buf[idx + 1..])
+                } else {
+                    (buf, false, &buf[buf.len()..])
+                };
+
+            if self.at_begining_of_line && !before_newline.is_empty() {
+                for _ in 0..self.level() {
+                    match self.indent {
```

Comment:
Its probably not a terrible perf hit doing it this way. If we find it is bad we can fix it like how you've described (though it would take a little extra tracking.

Reply:
It should be an easier call once the APIs `set_level`, `inc_level_by`, `dec_level_by`... are minimized.